src/app/plugins/platform/JavaScript/templates/webpack.config.js

```diff
@@ -28,226 +28,230 @@ const useCSSModules = false;
 }
 
-function createChunkFilenameTemplate(extension) {
+function createChunkFilenameTemplate(chunkFilenameTemplate, extension) {
   return chunkFilenameTemplate + '.' + extension;
 }
 
-// NB. Changing the name of the target directory requires updates to Capistrano and the
-// .gitignore file.
-const targetDirectory = path.join(__dirname, 'public');
-
-// We always use these plugins
-const plugins = [
-  // Remove stale build output
-  new CleanWebpackPlugin(),
-
-  // Create separate CSS stylesheets
-  new ExtractCssChunksPlugin({
-    filename: createFilenameTemplate('css'),
-    chunkFilename: createChunkFilenameTemplate('css'),
-  }),
-
-  // Type-check and run TSLint on all files in the project
-  new ForkTsCheckerPlugin({
-    async: false,
-    formatter: 'codeframe',
-    tslint: true,
-  }),
-
-  // Autogenerate a index.html for SPAs and webpack-dev-server.
-  new HtmlWebpackPlugin({
-    template: path.join(__dirname, 'src/index.html'),
-  }),
-];
-
 // Production plugins
 // If you're used to Webpack <= 3, you're probably wondering where we add production plugins.
 // The short answer is: we don't! Webpack 4 has built-in optimization options, and we're leaning on
 // its built-in defaults rather than overriding them.
 
-module.exports = {
-  context: __dirname,
+module.exports = env => {
+  const isProduction = env.NODE_ENV === 'production';
+  const isDevelopment = !isProduction;
 
-  mode: isProduction ? 'production' : 'development',
+// Derived settings from variables above:
+
+  const templatePrefix = isProduction ? '[id]' : '[name]';
+  const templateSuffix = hashFilenames ? '-[chunkhash]' : '';
+
+  const filenameTemplate = '[name]' + templateSuffix;
+  const chunkFilenameTemplate = templatePrefix + templateSuffix;
+  const imageFilenameTemplate = isProduction
+    ? '[hash].[ext]'
+    : '[name]-[hash].[ext]';
+
+  // NB. Changing the name of the target directory requires updates to
+  // Capistrano and the .gitignore file.
+  const targetDirectory = path.join(__dirname, 'public');
 
-  // Webpack defaults to 'eval' here, but that doesn't do a good job of preserving the TypeScript
-  // source maps output by ts-loader, so we use the slower source-map method.
-  devtool: isDevelopment && 'source-map',
+// We always use these plugins
+  const plugins = [
+    // Remove stale build output
+    new CleanWebpackPlugin(),
+
+    // Create separate CSS stylesheets
+    new ExtractCssChunksPlugin({
+      filename: createFilenameTemplate(filenameTemplate, 'css'),
+      chunkFilename: createChunkFilenameTemplate(chunkFilenameTemplate,'css'),
+    }),
+
+    // Type-check and run TSLint on all files in the project
+    new ForkTsCheckerPlugin({
+      async: false,
+      formatter: 'codeframe',
+      tslint: true,
+    }),
+
+    // Autogenerate a index.html for SPAs and webpack-dev-server.
+    new HtmlWebpackPlugin({
+      template: path.join(__dirname, 'src/index.html'),
+    }),
+  ];
+  return {
```

Comment:
Style nit: Prefer a new line before `return`ing.
```suggestion

  return {
```